File: scripts/update_research_data.py
"""
Script to update research data by cloning repositories, generating data, and processing it.
"""
import os
import shutil
import subprocess
import gzip
import sys
import argparse

# Constants
TEMP_DIR = os.path.abspath("temp_research_repos")
PUBLIC_DIR = os.path.abspath("public")
SCRIPTS_DIR = os.path.abspath("scripts")
VENV_DIR = os.path.join(TEMP_DIR, "venv")

REPOS = {
    "zsharp": "https://github.com/bangyen/zsharp.git",
    "oligopoly": "https://github.com/bangyen/oligopoly.git",
}


def run_cmd(cmd, cwd=None, env=None):
    """Executes a shell command."""
    print(
        f"Running: {' '.join(cmd) if isinstance(cmd, list) else cmd} in {cwd or os.getcwd()}"
    )
    subprocess.check_call(cmd, shell=isinstance(cmd, str), cwd=cwd, env=env)


def setup_venv():
    """Creates and configures the virtual environment."""
    if not os.path.exists(VENV_DIR):
        print("Creating virtual environment...")
        subprocess.check_call([sys.executable, "-m", "venv", VENV_DIR])
    else:
        print("Using existing virtual environment...")

    if sys.platform == "win32":
        pip_cmd = os.path.join(VENV_DIR, "Scripts", "pip")
        python_cmd = os.path.join(VENV_DIR, "Scripts", "python")
    else:
        pip_cmd = os.path.join(VENV_DIR, "bin", "pip")
        python_cmd = os.path.join(VENV_DIR, "bin", "python")

    return pip_cmd, python_cmd


def update_repos(args):
    """Clones or updates repositories."""
    if not args.skip_research:
        for name, url in REPOS.items():
            if name == "zsharp" and args.skip_zsharp:
                continue
            if name == "oligopoly" and args.skip_oligopoly:
                continue
            repo_dir = os.path.join(TEMP_DIR, name)
            if os.path.exists(repo_dir):
                print(f"Updating {name}...")
                run_cmd("git fetch origin", cwd=repo_dir)
                run_cmd("git reset --hard origin/main", cwd=repo_dir)
            else:
                print(f"Cloning {name}...")
                run_cmd(f"git clone {url} {name}", cwd=TEMP_DIR)


def install_dependencies(args, pip_cmd):
    """Installs dependencies for repositories."""
    if not args.skip_research:
        if not args.skip_zsharp:
            print("Installing ZSharp dependencies...")
            run_cmd([pip_cmd, "install", "."], cwd=os.path.join(TEMP_DIR, "zsharp"))

        if not args.skip_oligopoly:
            print("Installing Oligopoly dependencies...")
            run_cmd([pip_cmd, "install", "."], cwd=os.path.join(TEMP_DIR, "oligopoly"))

    print("Installing common dependencies (requests)...")
    run_cmd([pip_cmd, "install", "requests"], cwd=TEMP_DIR)


def run_zsharp_generation(args, python_cmd):
    """Runs ZSharp data generation."""
    if not args.skip_research and not args.skip_zsharp:
        # Copy script
        shutil.copy(
            os.path.join(SCRIPTS_DIR, "generate_zsharp_data.py"),
            os.path.join(TEMP_DIR, "zsharp", "generate_data.py"),
        )
        print("Generating ZSharp Data...")
        zsharp_cwd = os.path.join(TEMP_DIR, "zsharp")
        env = os.environ.copy()
        env["PYTHONPATH"] = zsharp_cwd
        run_cmd([python_cmd, "generate_data.py"], cwd=zsharp_cwd, env=env)


def run_oligopoly_generation(args, python_cmd):
    """Runs Oligopoly data generation."""
    if not args.skip_research and not args.skip_oligopoly:
        # Copy Script
        shutil.copy(
            os.path.join(SCRIPTS_DIR, "generate_oligopoly_matrix.py"),
            os.path.join(TEMP_DIR, "oligopoly", "generate_matrix.py"),
        )
        print("Generating Oligopoly Data...")
        oligopoly_cwd = os.path.join(TEMP_DIR, "oligopoly")
        env = os.environ.copy()
        env["PYTHONPATH"] = oligopoly_cwd
        run_cmd([python_cmd, "generate_matrix.py"], cwd=oligopoly_cwd, env=env)


def compress_output(args):
    """Compresses and moves output files."""
    if args.skip_research:
        return

    output_files = []
    if not args.skip_zsharp:
        zsharp_cwd = os.path.join(TEMP_DIR, "zsharp")
        output_files.append((
            os.path.join(zsharp_cwd, "zsharp_data.json"),
            os.path.join(PUBLIC_DIR, "zsharp_data.json.gz"),
        ))
    if not args.skip_oligopoly:
        oligopoly_cwd = os.path.join(TEMP_DIR, "oligopoly")
        output_files.append((
            os.path.join(oligopoly_cwd, "oligopoly_matrix.json"),
            os.path.join(PUBLIC_DIR, "oligopoly_data.json.gz"),
        ))

    for src, dst in output_files:
        if os.path.exists(src):
            print(f"Compressing {src} -> {dst}")
            with open(src, "rb") as f_in:
                with gzip.open(dst, "wb") as f_out:
                    shutil.copyfileobj(f_in, f_out)
        else:
            print(f"Error: Output file {src} not found!")


def main():
    """Main entry point."""
    parser = argparse.ArgumentParser(description="Update research data.")
    parser.add_argument("--clean", action="store_true", help="Clean temp directory.")
    parser.add_argument("--quizzes", action="store_true", help="Scrape quiz data.")
    parser.add_argument("--skip-research", action="store_true", help="Skip research data.")
    parser.add_argument("--skip-zsharp", action="store_true", help="Skip ZSharp.")
    parser.add_argument("--skip-oligopoly", action="store_true", help="Skip Oligopoly.")
    args = parser.parse_args()

    # 1. Clean and Setup Temp Dir
    if args.clean and os.path.exists(TEMP_DIR):
        print(f"Cleaning existing {TEMP_DIR}...")
        shutil.rmtree(TEMP_DIR)
    os.makedirs(TEMP_DIR, exist_ok=True)

    # 2. Setup Venv
    pip_cmd, python_cmd = setup_venv()

    # 3. Clone/Update
    update_repos(args)

    # 4. Install Deps
    install_dependencies(args, pip_cmd)

    # 5. Run Generations
    run_zsharp_generation(args, python_cmd)
    run_oligopoly_generation(args, python_cmd)

    # 6. Run Quiz Scraper
    if args.quizzes:
        print("Scraping Quizzes...")
        env = os.environ.copy()
        scrape_script = os.path.join(SCRIPTS_DIR, "scrape_quizzes.py")
        run_cmd([python_cmd, scrape_script], cwd=SCRIPTS_DIR, env=env)

    # 7. Compress
    compress_output(args)

    # 8. Cleanup
    if args.clean:
        print(f"Cleaning up {TEMP_DIR}...")
        shutil.rmtree(TEMP_DIR)
    else:
        print(f"Keeping {TEMP_DIR} for caching. Use --clean to remove.")

    print("Done! Research data updated in public/.")


if __name__ == "__main__":
    main()

File: scripts/updateResearchData.ts

 
import { execSync } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import zlib from 'node:zlib';

const TEMP_DIR = path.resolve(process.cwd(), 'temp_research_repos');
const PUBLIC_DIR = path.resolve(process.cwd(), 'public');
const VENV_DIR = path.join(TEMP_DIR, 'venv');
const IS_WIN = process.platform === 'win32';

const REPOS = {
    zsharp: 'https://github.com/bangyen/zsharp.git',
    oligopoly: 'https://github.com/bangyen/oligopoly.git',
};



// Helper functions
function runCmd(cmd: string, cwd: string = process.cwd(), env: NodeJS.ProcessEnv = process.env) {
    console.log(`Running: ${cmd} in ${cwd}`);
    execSync(cmd, { cwd, env, stdio: 'inherit' });
}

function setupVenv(): { pipCmd: string; pythonCmd: string } {
    if (fs.existsSync(VENV_DIR)) {
        console.log('Using existing virtual environment...');
    } else {
        console.log('Creating virtual environment...');
        execSync(`python3 -m venv ${VENV_DIR}`, { stdio: 'inherit' });
    }

    const binDir = IS_WIN ? 'Scripts' : 'bin';
    const pipCmd = path.join(VENV_DIR, binDir, 'pip');
    const pythonCmd = path.join(VENV_DIR, binDir, 'python');
    return { pipCmd, pythonCmd };
}

function updateRepos() {
    for (const [name, url] of Object.entries(REPOS)) {
        const repoDir = path.join(TEMP_DIR, name);
        if (fs.existsSync(repoDir)) {
            console.log(`Updating ${name}...`);
            runCmd('git fetch origin && git reset --hard origin/main', repoDir);
        } else {
            console.log(`Cloning ${name}...`);
            runCmd(`git clone ${url} ${name}`, TEMP_DIR);
        }
    }
}

function installDeps(pipCmd: string) {
    // Install repo deps
    for (const name of Object.keys(REPOS)) {
        const repoDir = path.join(TEMP_DIR, name);
        console.log(`Installing ${name} dependencies...`);
        runCmd(`${pipCmd} install .`, repoDir);
    }
    // Install common deps
    // console.log("Installing common dependencies (requests)...");
    // runCmd(`${pipCmd} install requests`, TEMP_DIR);
}

function runZSharp(pythonCmd: string) {
    const cwd = path.join(TEMP_DIR, 'zsharp');
    const scriptPath = path.join(cwd, 'generate_data.py');
    const sourcePath = path.join(process.cwd(), 'scripts/python/generate_zsharp_data.py');

    fs.copyFileSync(sourcePath, scriptPath);

    console.log('Running ZSharp Generation...');
    const env = { ...process.env, PYTHONPATH: cwd };
    runCmd(`${pythonCmd} generate_data.py`, cwd, env);
}

function runOligopoly(pythonCmd: string) {
    const cwd = path.join(TEMP_DIR, 'oligopoly');
    const scriptPath = path.join(cwd, 'generate_matrix.py');
    const sourcePath = path.join(process.cwd(), 'scripts/python/generate_oligopoly_matrix.py');

    fs.copyFileSync(sourcePath, scriptPath);

    console.log('Running Oligopoly Generation...');
    const env = { ...process.env, PYTHONPATH: cwd };
    runCmd(`${pythonCmd} generate_matrix.py`, cwd, env);
}

function compressAndMove() {
    // ZSharp
    const zSrc = path.join(TEMP_DIR, 'zsharp', 'zsharp_data.json');
    const zDst = path.join(PUBLIC_DIR, 'zsharp_data.json.gz');
    if (fs.existsSync(zSrc)) {
        console.log(`Compressing ${zSrc} -> ${zDst}`);
        const content = fs.readFileSync(zSrc);
        const compressed = zlib.gzipSync(content);
        fs.writeFileSync(zDst, compressed);
    } else {
        console.warn(`Warning: ${zSrc} not found.`);
    }

    // Oligopoly
    const oSrc = path.join(TEMP_DIR, 'oligopoly', 'oligopoly_matrix.json');
    const oDst = path.join(PUBLIC_DIR, 'oligopoly_data.json.gz');
    if (fs.existsSync(oSrc)) {
        console.log(`Compressing ${oSrc} -> ${oDst}`);
        const content = fs.readFileSync(oSrc);
        const compressed = zlib.gzipSync(content);
        fs.writeFileSync(oDst, compressed);
    } else {
        console.warn(`Warning: ${oSrc} not found.`);
    }
}

// Main
function main() {
    const args = process.argv.slice(2);
    const clean = args.includes('--clean');

    if (clean && fs.existsSync(TEMP_DIR)) {
        console.log(`Cleaning ${TEMP_DIR}...`);
        fs.rmSync(TEMP_DIR, { recursive: true, force: true });
    }
    if (!fs.existsSync(TEMP_DIR)) {
        fs.mkdirSync(TEMP_DIR, { recursive: true });
    }

    const { pipCmd, pythonCmd } = setupVenv();

    updateRepos();
    installDeps(pipCmd);

    runZSharp(pythonCmd);
    runOligopoly(pythonCmd);

    compressAndMove();

    if (clean) {
        console.log(`Cleaning up ${TEMP_DIR}...`);
        fs.rmSync(TEMP_DIR, { recursive: true, force: true });
    } else {
        console.log(`Keeping ${TEMP_DIR}. Use --clean to remove.`);
    }
    console.log('Done!');
}

main();
